qt/TorrentFilter.cc

```diff
@@ -83,103 +83,109 @@ int compare(T const a, T const b)
 
 } // namespace
 
+int TorrentFilter::lessThanState(Torrent const* left, Torrent const* right) const
+{
+    int val = compare(left->hasError(), right->hasError());
+
+    if (val == 0)
+    {
+        val = compare(left->isFinished(), right->isFinished());
+    }
+
+    if (val == 0)
+    {
+        val = -compare(left->isPaused(), right->isPaused());
+    }
+
+    if (val == 0)
+    {
+        val = -compare(left->isMagnet(), right->isMagnet());
+    }
+
+    return val;
+}
+
 bool TorrentFilter::lessThan(QModelIndex const& left, QModelIndex const& right) const
 {
-    int val = 0;
     Torrent const* a = sourceModel()->data(left, TorrentModel::TorrentRole).value<Torrent const*>();
     Torrent const* b = sourceModel()->data(right, TorrentModel::TorrentRole).value<Torrent const*>();
 
+    int val = lessThanState(a, b);
```

Comment:
Result of this call is ignored in half of the switch cases down below (with `val` assigned new value without checking its current value), meaning it needs to be called elsewhere, or in some other (delayed) way.